producer-kafka/src/main/java/io/stargate/producer/kafka/schema/SchemaRegistryProvider.java

```diff
@@ -0,0 +1,235 @@
+    return getLatestSchemaBySubject(subjectName);
+  }
+
+  @Override
+  public void createOrUpdateSchema(TableMetadata tableMetadata) {
+    createOrUpdateKeySchema(tableMetadata);
+    createOrUpdateValueSchema(tableMetadata);
+  }
+
+  private Schema getLatestSchemaBySubject(String subjectName) {
+    int latestSchemaId = -1;
+    try {
+      // be careful - possible race condition, if between getLatestSchemaMetadata and
+      // getSchemaBySubjectAndId the createOrUpdateSchema is called
+      latestSchemaId = schemaRegistryClient.getLatestSchemaMetadata(subjectName).getId();
+      return (Schema)
+          schemaRegistryClient.getSchemaBySubjectAndId(subjectName, latestSchemaId).rawSchema();
+    } catch (IOException | RestClientException e) {
+      throw new RuntimeException(
+          "Problem when get key schema for subject: "
+              + subjectName
+              + " and schema id: "
+              + latestSchemaId,
+          e);
+    }
+  }
+
+  private void createOrUpdateValueSchema(TableMetadata tableMetadata) {
+    ParsedSchema valueSchema = new AvroSchema(constructValueSchema(tableMetadata));
+    String subject = constructValueRecordName(tableMetadata);
+
+    int schemaId = registerSchema(tableMetadata, valueSchema, subject);
+
+    logger.info(
+        "Registered valueSchema: {}, for subject: {} and id: {}", valueSchema, subject, schemaId);
+  }
+
+  private void createOrUpdateKeySchema(TableMetadata tableMetadata) {
+    ParsedSchema keySchema = new AvroSchema(constructKeySchema(tableMetadata));
+    String subject = constructKeyRecordName(tableMetadata);
+
+    int schemaId = registerSchema(tableMetadata, keySchema, subject);
+
+    logger.info(
+        "Registered keySchema: {}, for subject: {} and id: {}", keySchema, subject, schemaId);
+  }
+
+  private int registerSchema(TableMetadata tableMetadata, ParsedSchema schema, String subject) {
+    try {
+      return schemaRegistryClient.register(subject, schema);
+    } catch (IOException | RestClientException e) {
+      throw new RuntimeException(
+          "Problem when create or update key schema for subject: " + subject, e);
+    }
+  }
+
+  Schema constructKeySchema(TableMetadata tableMetadata) {
+    String keyRecordName = constructKeyRecordName(tableMetadata);
+    FieldAssembler<Schema> keyBuilder = SchemaBuilder.record(keyRecordName).fields();
+
+    for (ColumnMetadata columnMetadata : tableMetadata.getPartitionKeys()) {
+      Schema avroFieldSchema = CqlToAvroTypeConverter.toAvroType(columnMetadata.getType());
+      keyBuilder.name(columnMetadata.getName()).type(avroFieldSchema).noDefault();
+    }
+    return keyBuilder.endRecord();
+  }
+
+  Schema constructValueSchema(TableMetadata tableMetadata) {
+    List<Schema> partitionKeys =
+        constructRequiredValueFieldsSchema(tableMetadata.getPartitionKeys());
+    List<Schema> clusteringKeys =
+        constructRequiredValueFieldsSchema(tableMetadata.getClusteringKeys());
```

Comment:
Clustering keys might or might not be present in a DELETE event.

Reply:
and that is perfectly ok, this `required` means not-null value, but the field can be missing because it is a union type:
`SchemaBuilder.unionOf().nullType().and().type(partitionKey).endUnion()`
so this:
`CK : null` will not be possible, but absence of the `CK` is.
I've renamed those methods, let me know if it is more clear now.

Reply:
oh, got it, thanks for the info.